### ISSUE TITLE
Cannot directly find match results in entity search anymore

### ISSUE THREAD

**Issue by emufan** [home-assistant/frontend#17989]
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

With another recent change, the entity search (keypress e, or add entity to card search list, dev tools entity state change entity picker, etc.) is now again a fuzzy search with not good prioritizations.

Early in 2022.5 this was introduced already once, but according to core team, user feedback, etc. reverted directly again.

See:
https://github.com/home-assistant/frontend/discussions/12590#discussioncomment-2722057
https://community.home-assistant.io/t/2022-5-streamlining-settings/418477/376

Now this is back again. Perhaps the fuzzy order is better (cannot test it) but it is not good at all or still the problem that was the reason to revert it directly in 2022.5.3 again.

If search for "cla" the first exact match in search result item 34 (!):

![image](https://github.com/home-assistant/frontend/assets/22775515/c6670975-373d-4063-aeea-cb630358468a)

If I search for "counter" I get first some cover entries and then the first exact match counter

![image](https://github.com/home-assistant/frontend/assets/22775515/012ee336-b4e0-4c4e-bf0e-82e19a78a96d)

If I search vor "laut" the first result I want to see is on position 4 and the rest of the exact matches containing "laut" is then randomly distributed in a list result list with >300 entries. So not even on result position 5, 6, 7. With this I cannot find the entities anymore containing "laut" at all. BTW I would doubt that a >300 random result list would help any user to find what he is searching for.

![image](https://github.com/home-assistant/frontend/assets/22775515/f2fb76f0-2a82-4edd-9c15-3e08bcee6810)

If and only if the complete hits/exact containing matches would be shown on top or with manual * or a fuzzy-checkbox it make sense for me. But not when I don’t find exact matches anymore in the list.

I would agree to see such result lists if I would be able to enter "l\*a\*u\*t". But wildcards are not supported here. But without wildcards the results have to show exact matches results only or every (!) exact matches results on top of the result list. Loop again over current result list and put exact containing matches on top of the current list. Then we would have perhaps both worlds, the old exact results and and wildcard results afterwards.

If “everyone” in the linked history agreed that it wasn’t a good idea and reverted is back, why should this now a better idea to re-do the same mistake again. Who is why re-putting already reverted “features” back?

### Describe the behavior you expected

Either:

a. Option without wildcards:

- show only exact matches
- show all (!) exact matches in result list on top of the list and then the rest of the fuzzy
- Have a checkbox (fuzzy-non-fuzzy) in the search

Option 2 is bar minimum. Loop again over current result list and put exact containing matches on top of the current list. Then we would have perhaps both worlds, the old exact results and and wildcard results afterwards.

b. Option with wildcard:

- Show only exact matches (with respecting the wildcards), so "cla" woud find only entities with existing "cla" and "c\*l\*a" would find everything accoding to the pattern, 

### Steps to reproduce the issue

Search for exact matches entity search (keypress e, or add entity to card search list, dev tools entity state change entity picker, etc.) and see that exact matches are not on top on the result list.

### What version of Home Assistant Core has the issue?

2023.9.2

### What was the last working version of Home Assistant Core?

2023.8

### In which browser are you experiencing the issue with?

all

### Which operating system are you using to run this browser?

Win

**Comment by ildar170975** [comment]
Just to link some recent Discussion: https://github.com/home-assistant/frontend/discussions/17474

Agreed, the current search is:
-- different in different places;
-- gives plenty of useless results.

**Comment by leeoniya** [comment]
most fuzzy match libs give useless results.

i've done a pretty thorough comparison (with live demos) when writing [uFuzzy](https://github.com/leeoniya/uFuzzy#a-biased-appraisal-of-similar-work)

**Comment by emufan** [comment]
Yes, but this is then a pro or a cons against these search results here in HA? For me a "it is as it is, everything is same wrong elsewhere" shouldn't be a pro for the implementation.

At least direct matches should be always and complete on top of the result list. And as said, I doubt that a result list with 300 entries has any benefit at all.

**Comment by emufan** [comment]
Still valid.

**Comment by emufan** [comment]
Still valid.

**Comment by emufan** [comment]
still valid

At least direct matches should be always and complete on top of the result list.

**Comment by emufan** [comment]
still valid

At least direct matches should be always and complete on top of the result list.

**Comment by emufan** [comment]
still valid

At least direct matches should be always and complete on top of the result list.

**Comment by ildar170975** [comment]
2025.6.1

![Image](https://github.com/user-attachments/assets/68a9c446-405a-4abd-a019-03257ca6557a)

**Comment by tomlut** [comment]
Yep the search results are terrible. Why is `pond level v` not the first item on the list?

<img width="1016" height="576" alt="Image" src="https://github.com/user-attachments/assets/b7c05b11-1722-42b7-9bed-9423fb23ac1e" />

**Comment by Leah974843** [comment]
I've been noticing the same thing recently. One of the problems is that it's searching for ANY of the words instead of all of them. When I'm searching for switch.living, I don't want switch.dining things. I either want switch.living... or switch.somethingelse_living.

<img width="1170" height="501" alt="Image" src="https://github.com/user-attachments/assets/8acad547-2cdf-4b0f-b580-b30c2766d210" />

**Comment by alexvaltchev** [comment]
I have previously reported this problem awhile back, and it was fixed withing 1-2 months. Now it has somehow come back again and is very annoying!!! https://community.home-assistant.io/t/home-assistant-device-entity-names-search-does-not-show-correctly-in-drop-downs-side-menus-now/926126/1

Can someone please fix this, I've noticed it again for the past 2-3 version AT LEAST, maybe more... If it was fixed once, I am sure it can be fixed again. Thank you!

Kinds regards,
Alex

**Comment by alexvaltchev** [comment]
NOTHING? Really? Not even a single post? This is a very serious problem that has been going on for month, why so many new things are implemented but a bug of this magnitude is not taken care of? That is beyond me... What is the point of creating new things when the old ones don't work the way they are indented? HA is an awesome software but for some reason i see trend to do as many new features every time, yet some of them are not even stable or buggy... Isn't stability the core of it? If something is not functional what good does it do?

**Comment by teskanoo** [comment]
I have 5000+ entities - it is a real chore to find entities, even providing the exact friendly name or the exact object or exact entity id cannot being the desired entry to the top of the list.

Furthermore, when trying to filter in the devices-and-services/entities grid, this is very problematic. 

In this interface I want exact matches so I can apply labels to entities without having 50+ matches to visually double-check to find the 8 items I am trying to find and apply labels to.

I wonder if we could at least have a 'no-fuzzy-logic' / 'exact-match' checkbox in the devices, entities, helpers, labels etc. interfaces so we can more easily manage our installations.

I'm sure the feature is very much appreciated by users with smaller installations, very helpful when you forget the name of something or are searching for the alias, label applied etc. But a power-user option would be much appreciated

**Comment by tomlut** [comment]
Still an issue

**Comment by teskanoo** [comment]
I had hoped that this 'fuzzy' situation would have been improved in the new 2026.2.1 release where the combined search dialog was implemented. An Ideal time to look at other issues related to search, but the situation remains the same. I will acknowledge that the 'fuzzy' search logic is less destructive in the 'user' facing search for dashboards, however, in the entities page I retain that this is inappropriate default behaviour in that it negates the usefulness of the Search->Select-All->Bulk-Update paradigm

**Comment by ildar170975** [comment]
Entity picker - same, matches are not on the top:

<img width="423" height="330" alt="Image" src="https://github.com/user-attachments/assets/691f3aa1-292b-4800-996f-1f758ecb7760" />

You need to input ".test_level" to get results:

<img width="394" height="286" alt="Image" src="https://github.com/user-attachments/assets/96309f53-3124-4e9b-bc2f-181da9e9ab6f" />

**Comment by alexvaltchev** [comment]
Same here! I HATE HATE HATE this!!! It has been broken for months, when will it finally get fixed!?!?!

**Comment by karwosts** [comment]
>  however, in the entities page I retain that this is inappropriate default behaviour in that it negates the usefulness of the Search->Select-All->Bulk-Update paradigm

@teskanoo I thought fuzzy search was deprioritized from data-tables search from this PR in January. Do you experience differently?

https://github.com/home-assistant/frontend/pull/28853

**Comment by teskanoo** [comment]
It is deprioritized,and closer matches do appear at the top of the filtered list.

 My point is that all the fuzzy results do allow a user to filter down to a list of ONLY what is wanted. 

The results always contain partial matches and the implied suggestion  that what I really meant to type was a word that is similar to the word I actually typed.

This breaks the formerly possible workflow of ...

- filter
- sect all
- apply labels or enable/disable.

Having more fuzzy results than exact matches means that in spite of typing a substring that is actually present in 21 entities, I have to manually click the checkbox on every one of these entities before applying bulk updates.

**Comment by karwosts** [comment]
Share an example screenshot? AFAIK when you have an exact substring match it's supposed to disable fuzzy searching. 

edit: this applies to data-tables only

**Comment by tomlut** [comment]
Clearly not:

<img width="529" height="530" alt="Image" src="https://github.com/user-attachments/assets/0e0af49f-521d-41c6-9990-2da9050b4885" />

**Comment by lichenophile** [comment]
Fuzzy search, where-ever it is implemented, has always been awful. In Home Assistant, it's one of the worst example cases of how terrible it can be. What once worked flawlessly, is now a hot mess. 

Instead of providing yet more examples for the massive pile already here, I'll suggest a solution I've seen elsewhere, that would solve all issues while also pleasing those dozen people out there that actually _like_ fuzzy search.

<img width="544" height="90" alt="Image" src="https://github.com/user-attachments/assets/7cedd8e3-d6f4-4028-8d8e-c5940a9c5b1e" />

<img width="550" height="87" alt="Image" src="https://github.com/user-attachments/assets/e3755b10-b0ed-415d-b1ad-9c81f4365474" />

Have it retain the chosen mode for the user.

**Comment by karwosts** [comment]
> Clearly not:

Sorry my comment on that behavior was _only_ referring to the ha-data-table (e.g. config/entities filtering), that behavior wasn't applied to search pickers.

**Comment by JonLaliberte** [comment]
I'd say it definitely doesn't return the options I'd deem the most relevant.

Example:

<img width="592" height="634" alt="Image" src="https://github.com/user-attachments/assets/322f5524-8fc5-43e9-aacf-134f07c29858" />

I think you could argue that returning sensors before switches makes sense. I'd then have to argue that you're wrong though.
